content/blog/2025-12-16-gvn-final-report.md

```diff
@@ -0,0 +1,1350 @@
+
+    return seq
+```
+
+#### 2.5.3 The Bug We Fixed
+
+The original algorithm would loop infinitely on certain copy patterns because it didn't correctly identify when to break a cycle. Specifically, it failed when:
+
+1. A destination was used as a source by another copy
+2. Multiple variables formed a cycle
+
+The fix correctly identifies cycle members (variables that are both sources and destinations) and breaks the cycle by saving one to a temporary.
+
+**Example of the fix in action:**
+```
+# Parallel: (x, y, z) := (y, z, x)  # 3-way cycle
+# Step 1: No safe copy (all dests are sources)
+# Step 2: Save x to tmp: tmp = x
+# Step 3: Update: (x, y, z) := (y, z, tmp)
+# Step 4: Now z := tmp is safe, emit it
+# Step 5: Now y := z is safe, emit it
+# Step 6: Now x := y is safe, emit it
+# Result: tmp=x; z=tmp; y=z; x=y
+```
+
+### 2.6 Real Phi Node Representation {#26-real-phi-node-representation}
+
+#### 2.6.1 Why Real Phi Nodes?
+
+The original GVN paper by Cooper and Briggs uses real phi nodes in SSA form. Bril provides an alternative approach using `get`/`set` operations with labels, which is easier to implement but doesn't faithfully represent SSA.
+
+We chose to implement **real phi nodes** for several reasons:
+
+1. **Algorithmic Fidelity**: The GVN algorithm naturally operates on phi nodes as first-class instructions
+2. **Value Numbering of Phis**: We can assign value numbers to phi nodes and detect redundant phis
+3. **Learning Value**: Understanding real SSA conversion is pedagogically important
+
+#### 2.6.2 Phi Node Structure
+
+Our phi nodes have the following structure:
+
+```json
+{
+  "op": "phi",
+  "dest": "x.3",
+  "type": "int",
+  "args": ["x.1", "x.2"],
+  "labels": ["B1", "B2"]
+}
+```
+
+The `args` and `labels` arrays are parallel — `args[i]` is the value when control comes from `labels[i]`.
+
+#### 2.6.3 Phi Node Processing in GVN
+
+Phi nodes are processed specially in the GVN walk:
+
+```python
+def process_phi(phi: dict, ctx: GVNContext, env: Dict[str, VN],
+                val2rep: Dict[Tuple, str], has_backedge: bool) -> VN:
+    dest = phi["dest"]
+    args = phi.get("args", [])
+
+    if has_backedge:
+        # Conservative: fresh VN for loop header phis
+        return ctx.new_vn()
+
+    # Map phi args to their VNs
+    arg_vns = [ctx.get_vn(arg).n for arg in args]
+
+    # Trivial phi: all args have same VN
+    if all(v == arg_vns[0] for v in arg_vns):
+        return VN(arg_vns[0])
+
+    # Check for redundant phi (same arg VN pattern)
+    phi_key = ("phi", type_key(ty), tuple(arg_vns))
+    existing = ctx.lookup_expr(phi_key)
+    if existing:
+        return existing
+
+    vn = ctx.new_vn()
+    ctx.insert_expr(phi_key, vn)
+    return vn
+```
+
+This allows GVN to:
+- Recognize trivial phis (all args same) and propagate the value
+- Detect redundant phis (same pattern of arg VNs)
+- Be conservative at loop headers
+
+### 2.7 Copy Coalescing {#27-copy-coalescing}
```

Comment:
This section is a VERY long way to introduce an extremely simple copy-prop pass… can you summarize it in a single sentence?